Remove validation from __init__ and set_params from SGDOneClassSVM

MrinalTyagi opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Address:  #21406

#### What does this implement/fix? Explain your changes.
Add type check of alpha in _validation in SGDOneClassSVM class

#### Any other comments?
Support from someone in case of error in merging will be highly appreciable.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

MrinalTyagi committed:
Base Version

MrinalTyagi committed:
Update _stochastic_gradient.py

MrinalTyagi committed:
Update _stochastic_gradient.py

MrinalTyagi committed:
Update _stochastic_gradient.py

MrinalTyagi committed:
Update _stochastic_gradient.py

MrinalTyagi committed:
Update _stochastic_gradient.py

MrinalTyagi commented:
@glemaitre Could you please guide how to proceed in the following issue? If I let alpha stay in init then it will fail but if I move alpha assignment to _validation_params it gives me a error of setting class attribute alpha is not acceptable.

ogrisel commented:
This will likely be complex to fix because of the class hierarchy that implies shared code while enforcing the same name for a reparametrized hyper-parameter...

+1 the backward compat handling that will be tricky to get right. @MrinalTyagi let me suggest to try to tackle another estimator. Getting this one right will open a can of worms.

MrinalTyagi commented:
@ogrisel Sure. I am willing to contributions to any type of issue or feature addition so you can definitely provide me another estimator as well as other issues if you feel right as I have just started working on open source and wanna continue it.

arisayosh commented:
Hi @MrinalTyagi and @ogrisel, 

@iofall and I started working on this last week not realizing that a solution had already been drafted. We did the following and it seemed to pass all of the test required: 

1. Changed `alpha` to None in `__init__`
2. Previously `alpha` was initialized as `self.nu / 2` in `__init__` which was performed before validation of `nu`. This caused type errors in the `__init__` method. 
3. Now `alpha` is initialized as `None` and modified after validation in its respective `fit()` & `partial_fit()` methods

Let me know what you think!

MrinalTyagi commented:
> Hi @MrinalTyagi and @ogrisel,
> 
> @iofall and I started working on this last week not realizing that a solution had already been drafted. We did the following and it seemed to pass all of the test required:
> 
> 1. Changed `alpha` to None in `__init__`
> 2. Previously `alpha` was initialized as `self.nu / 2` in `__init__` which was performed before validation of `nu`. This caused type errors in the `__init__` method.
> 3. Now `alpha` is initialized as `None` and modified after validation in its respective `fit()` & `partial_fit()` methods
> 
> Let me know what you think!

Test like the build test or the individual test for this class that runs using `pytest sklearn/tests/test_common.py::test_estimators_do_not_raise_errors_in_init_or_set_params
`

ogrisel commented:
@arisayosh I am not sure I understand the details of your plan because when you speak of `__init__` you do not precise which class you are referring to, but it's possible that your strategy can work (but the devil is in the details). Feel free to open a concurrent PR.

arisayosh commented:
> Test like the build test or the individual test for this class that runs using `pytest sklearn/tests/test_common.py::test_estimators_do_not_raise_errors_in_init_or_set_params `

Hi @MrinalTyagi, we tested `pytest sklearn/tests/test_common.py::test_estimators_do_not_raise_errors_in_init_or_set_params` and `pytest sklearn/linear_model/tests .`

We've opened a concurrent PR here for these changes --> #21944. The `alpha` is set to `None` during instantiation of  `SGDOneClassSVM` and set to `self.nu/2` during `fit()` and `partial_fit()` before `self._validate_params()`.

cc: @ogrisel, @iofall

jeremiedbb commented:
Done in https://github.com/scikit-learn/scikit-learn/pull/24433
